iree/compiler/Dialect/VM/Transforms/DeduplicateRodata.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+#include "iree/compiler/Dialect/VM/IR/VMDialect.h"
+#include "iree/compiler/Dialect/VM/IR/VMOps.h"
+#include "iree/compiler/Dialect/VM/IR/VMTypes.h"
+#include "iree/compiler/Dialect/VM/Transforms/Passes.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace VM {
+
+class DeduplicateRodataPass
+    : public PassWrapper<DeduplicateRodataPass,
+                         OperationPass<IREE::VM::ModuleOp>> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::VM::VMDialect>();
+  }
+
+  StringRef getArgument() const override {
+    return "iree-vm-deduplicate-rodata";
+  }
+
+  StringRef getDescription() const override {
+    return "Deduplicates vm.rodata ops in the module.";
+  }
+
+  using RodataKey = std::tuple<StringRef, Attribute>;
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+
+    // Gather all rodata ops with the same value.
+    DenseMap<RodataKey, SmallVector<IREE::VM::RodataOp>> bucketedOps;
+    for (auto rodataOp : moduleOp.getOps<IREE::VM::RodataOp>()) {
+      if (rodataOp.ordinal().hasValue()) {
+        rodataOp.emitError() << "rodata op already has an ordinal assigned; "
+                                "cannot perform deduplication";
+        return signalPassFailure();
+      }
+      RodataKey key = std::make_tuple(rodataOp.mime_type().getValueOr(""),
+                                      rodataOp.value());
```

Comment:
How large a memory footprint do these keys have for large rodata values? (I'm worried about OOMing host machines when compiling large models if we keep multiple copies of large buffers in memory, hashing or storing references could help if that isn't happening already)

Reply:
All attributes are uniqued, so just the size of two pointers.